### PR TITLE
CI: split nested GHA groups instead of panicking

### PR DESCRIPTION
**Pull request by Kobzol**
Bootstrap uses Github Actions groups to reduce clutter in CI job output. However, GHA doesn't support group nesting, and currently, when a group would be nested, bootstrap would panic. This is causing intermittent CI failures, because it's not trivial to make sure that groups won't be nested, and subtle changes in bootstrap (or even in caches being present) can cause nesting.

This PR changes the logic so that groups are never nested. Instead, when a group would be nested, the previous group is ended, and only then is the subgroup started. When the subgroup finishes, it will then restart any previously ended parent group.

r? @Mark-Simulacrum

### CODE REVIEW

File: src/tools/build_helper/src/ci.rs [hunk 1]
```diff
@@ -36,25 +36,26 @@ impl CiEnv {
 }
 
 pub mod gha {
-    use std::sync::atomic::{AtomicBool, Ordering};
+    use std::sync::{Mutex, OnceLock};
 
-    static GROUP_ACTIVE: AtomicBool = AtomicBool::new(false);
+    static ACTIVE_GROUPS: OnceLock<Mutex<Vec<String>>> = OnceLock::new();
```
Comment (Mark-Simulacrum) [inline src/tools/build_helper/src/ci.rs:41]: Do we need a OnceLock here? [Mutex::new](https://doc.rust-lang.org/nightly/std/sync/struct.Mutex.html#method.new) is const?

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me with nit fixed

File: src/tools/build_helper/src/ci.rs (hunk 1 above)
Comment (Kobzol) [reply src/tools/build_helper/src/ci.rs]: Good point, thanks!

**Pull request merged**